Data driven CSS property parsing

weinig opened the pull request:
#### 33e16a60b9c5713b5df84d9b59f212e36bb1e3ce
<pre>
Data driven CSS property parsing
<a href="https://bugs.webkit.org/show_bug.cgi?id=248036">https://bugs.webkit.org/show_bug.cgi?id=248036</a>
rdar://102467693

Reviewed by Darin Adler.

Introduce the initial infrastructure for data/grammar driven
parsing of CSS properties, starting with parsing of keyword
only property values.

The long term goal is to add enough information in CSSProperties.json
to generate the property parsers and ASTs, but this initial
change just puts in place enough to generate the boilerplate
for CSSPropertyParser::parseSingleValue() (e.g. we now generate
the big switch statement) with most of the parsing still being
done by hand written &quot;consume&quot; functions. To make inclusion
easier and simplify where to look for the consume functions,
the existing consume functions in CSSPropertyParser.cpp were
moved to CSSPropertyParserHelpers, where most of the other
consume functions already lived.

In addition to generating the core switch statement, this change
also generates the consume functions for properties that support
taking a single keyword CSSValue as their input. This utilized
the existing &quot;values&quot; arrays CSSProperties.json, updated as needed
to match the current status in WebCore. The same data is also
used to generate implementations for the CSSParserFastPaths&apos;
`isValidKeywordPropertyAndValue` and `isKeywordPropertyID`. In
the future, more properties that allow single keywords can be
automatically added based on their grammar, with no additional
work needed by the person adding them.

To aid the transition to generated parsing, a number of new
attributes were added to the &quot;codegen-properties&quot; dictionaries
in CSSProperties.json for specifying custom consume functions
and optionally passing additional arguments.

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
Added new files.

* Source/WebCore/css/CSSFontVariantLigaturesParser.h: Added.
* Source/WebCore/css/CSSFontVariantNumericParser.h: Added.
Moved here from CSSPropertyParser.cpp so it can be used in multiple files.

* Source/WebCore/css/CSSProperties.json:
Update properties to indicate where a custom parser is needed and if any
additional parameterization of that custom parser is required. Also updates
the &quot;values&quot; arrays where they were out of sync with the engine as they are
now used if no custom parser is requested.

* Source/WebCore/css/CSSTimingFunctionValue.cpp:
Add a missing newline.

* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::ComputedStyleExtractor::propertyValue):
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle):
Use the new reference overload of isExposed to avoid a null check.

* Source/WebCore/css/parser/CSSAtRuleID.h:
Add a missing newline.

* Source/WebCore/css/parser/CSSParserFastPaths.h:
* Source/WebCore/css/parser/CSSParserFastPaths.cpp:
(WebCore::CSSParserFastPaths::isValidKeywordPropertyAndValue):
(WebCore::CSSParserFastPaths::isKeywordPropertyID):
Replace implementations with calls to generated versions.

(WebCore::CSSParserFastPaths::isPartialKeywordPropertyID): Deleted.
No longer needed, as the one caller now can handle this in
generated code.

* Source/WebCore/css/parser/CSSParserIdioms.cpp:
Add a missing newline.

* Source/WebCore/css/parser/CSSParserImpl.cpp:
(WebCore::CSSParserImpl::CSSParserImpl):
Remove an extraneous newline and fix an includes&apos;s style.

* Source/WebCore/css/parser/CSSPropertyParser.cpp:
* Source/WebCore/css/parser/CSSPropertyParser.h:
Move lots of things out. Generate parseSingleValue.

* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
* Source/WebCore/css/parser/CSSPropertyParserHelpers.h:
Move all the loose consume functions in from CSSPropertyParser.cpp
so they can used in multiple files (including the new generated ones).

* Source/WebCore/css/process-css-properties.py:
Generates new files CSSPropertyParsing.h/cpp which contain the base
switch for parsing single longhand values as well as the keyword only
checking and consumer functions.

* Source/WebCore/css/typedom/StylePropertyMap.cpp:
(WebCore::StylePropertyMap::set):
(WebCore::StylePropertyMap::remove):
Use the new reference overload of isExposed to avoid a null check.

Canonical link: <a href="https://commits.webkit.org/256867@main">https://commits.webkit.org/256867@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b199d58096b49db39b4643f813ada6cbee3d1198

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/97040 "7 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/6308 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/30155 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/106561 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/166835 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/101013 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/6538 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/35039 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/89432 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/103254 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/102710 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/4912 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/83656 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/31923 "Passed tests") | 
| [❌ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/5/builds/100628 "Found 1 webkitpy python3 test failure: webkitcorepy.tests.file_lock_unittest.FileLockTestCase.test_locked_timeout") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/86755 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/88623 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/74814 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/338 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/20102 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/321 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/21529 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4750 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/5114 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/44038 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/1557 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/40836 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/css/CSSFontVariantLigaturesParser.h:
```diff
@@ -0,0 +1,103 @@
+        : m_sawCommonLigaturesValue(false)
+        , m_sawDiscretionaryLigaturesValue(false)
+        , m_sawHistoricalLigaturesValue(false)
+        , m_sawContextualLigaturesValue(false)
```
Can we initialize these below instead?

darinadler commented on Source/WebCore/css/CSSFontVariantNumericParser.h:
```diff
@@ -0,0 +1,109 @@
+        , m_sawNumericSpacingValue(false)
+        , m_sawNumericFractionValue(false)
+        , m_sawOrdinalValue(false)
+        , m_sawSlashedZeroValue(false)
```
Can we initialize these below instead?

darinadler commented on Source/WebCore/css/CSSFontVariantNumericParser.h:
```diff
@@ -0,0 +1,109 @@
+        ConsumedValue,
+        DisallowedValue,
+        UnknownValue
+    };
```
Can we add `: uint8_t`, and also do this as a one-liner?

darinadler commented on Source/WebCore/css/CSSFontVariantLigaturesParser.h:
```diff
@@ -0,0 +1,103 @@
+        ConsumedValue,
+        DisallowedValue,
+        UnknownValue
+    };
```
Can we add `: uint8_t`, and also do this as a one-liner?

Commit:
Data driven CSS property parsing
https://bugs.webkit.org/show_bug.cgi?id=248036
rdar://102467693

Reviewed by Darin Adler.

Introduce the initial infrastructure for data/grammar driven
parsing of CSS properties, starting with parsing of keyword
only property values.

The long term goal is to add enough information in CSSProperties.json
to generate the property parsers and ASTs, but this initial
change just puts in place enough to generate the boilerplate
for CSSPropertyParser::parseSingleValue() (e.g. we now generate
the big switch statement) with most of the parsing still being
done by hand written "consume" functions. To make inclusion
easier and simplify where to look for the consume functions,
the existing consume functions in CSSPropertyParser.cpp were
moved to CSSPropertyParserHelpers, where most of the other
consume functions already lived.

In addition to generating the core switch statement, this change
also generates the consume functions for properties that support
taking a single keyword CSSValue as their input. This utilized
the existing "values" arrays CSSProperties.json, updated as needed
to match the current status in WebCore. The same data is also
used to generate implementations for the CSSParserFastPaths'
`isValidKeywordPropertyAndValue` and `isKeywordPropertyID`. In
the future, more properties that allow single keywords can be
automatically added based on their grammar, with no additional
work needed by the person adding them.

To aid the transition to generated parsing, a number of new
attributes were added to the "codegen-properties" dictionaries
in CSSProperties.json for specifying custom consume functions
and optionally passing additional arguments.

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
Added new files.

* Source/WebCore/css/CSSFontVariantLigaturesParser.h: Added.
* Source/WebCore/css/CSSFontVariantNumericParser.h: Added.
Moved here from CSSPropertyParser.cpp so it can be used in multiple files.

* Source/WebCore/css/CSSProperties.json:
Update properties to indicate where a custom parser is needed and if any
additional parameterization of that custom parser is required. Also updates
the "values" arrays where they were out of sync with the engine as they are
now used if no custom parser is requested.

* Source/WebCore/css/CSSTimingFunctionValue.cpp:
Add a missing newline.

* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::ComputedStyleExtractor::propertyValue):
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle):
Use the new reference overload of isExposed to avoid a null check.

* Source/WebCore/css/parser/CSSAtRuleID.h:
Add a missing newline.

* Source/WebCore/css/parser/CSSParserFastPaths.h:
* Source/WebCore/css/parser/CSSParserFastPaths.cpp:
(WebCore::CSSParserFastPaths::isValidKeywordPropertyAndValue):
(WebCore::CSSParserFastPaths::isKeywordPropertyID):
Replace implementations with calls to generated versions.

(WebCore::CSSParserFastPaths::isPartialKeywordPropertyID): Deleted.
No longer needed, as the one caller now can handle this in
generated code.

* Source/WebCore/css/parser/CSSParserIdioms.cpp:
Add a missing newline.

* Source/WebCore/css/parser/CSSParserImpl.cpp:
(WebCore::CSSParserImpl::CSSParserImpl):
Remove an extraneous newline and fix an includes's style.

* Source/WebCore/css/parser/CSSPropertyParser.cpp:
* Source/WebCore/css/parser/CSSPropertyParser.h:
Move lots of things out. Generate parseSingleValue.

* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
* Source/WebCore/css/parser/CSSPropertyParserHelpers.h:
Move all the loose consume functions in from CSSPropertyParser.cpp
so they can used in multiple files (including the new generated ones).

* Source/WebCore/css/process-css-properties.py:
Generates new files CSSPropertyParsing.h/cpp which contain the base
switch for parsing single longhand values as well as the keyword only
checking and consumer functions.

* Source/WebCore/css/typedom/StylePropertyMap.cpp:
(WebCore::StylePropertyMap::set):
(WebCore::StylePropertyMap::remove):
Use the new reference overload of isExposed to avoid a null check.

Canonical link: https://commits.webkit.org/256867@main
